server/src/main/java/org/elasticsearch/search/fetch/FetchContext.java

```diff
@@ -36,16 +37,53 @@
 public class FetchContext {
 
     private final SearchContext searchContext;
-    private final SearchLookup searchLookup;
     private final SourceLoader sourceLoader;
+    private final FetchSourceContext fetchSourceContext;
+    private final StoredFieldsContext storedFieldsContext;
 
     /**
      * Create a FetchContext based on a SearchContext
      */
     public FetchContext(SearchContext searchContext) {
         this.searchContext = searchContext;
-        this.searchLookup = searchContext.getSearchExecutionContext().lookup();
         this.sourceLoader = searchContext.newSourceLoader();
+        this.storedFieldsContext = buildStoredFieldsContext(searchContext);
+        this.fetchSourceContext = buildFetchSourceContext(searchContext);
+    }
+
+    private static FetchSourceContext buildFetchSourceContext(SearchContext in) {
+        FetchSourceContext fsc = in.fetchSourceContext();
+        StoredFieldsContext sfc = in.storedFieldsContext();
+        if (fsc == null) {
+            boolean hasStoredFields = in.hasStoredFields();
+            boolean hasScriptFields = in.hasScriptFields();
+            // TODO it seems a bit odd that we disable implicit source loading if we've asked
+            // for stored fields or script fields? But not eg doc_value fields or via
+            // the `fields` API
+            if (hasStoredFields == false && hasScriptFields == false) {
```

Comment:
Moving all this logic into one place has made it clear that we have some inconsistencies over how we decide when to load source which we should probably clear up.  Currently we load source unless you specifically ask us not to, *or* if you don't mention source at all in the request but do request stored or script fields.  This seems a bit odd?